### ISSUE TITLE
The results of an Aggregation (ResultsIterator<E>) can only be iterated over once.

### ISSUE THREAD

**Issue by eric-2d** [bguerout/jongo#263]
Jongo v1.2

When performing an aggregation like so:

`Iterable<MyObject> res = collection.aggregate(query).as(MyObject.class);`

res is only iterable once.
That's because ResultsIterator<E>.iterator() returns this. So if this iterator is already at the end, there is no way to obtain a new one pointing to the beginning.

As a comparison, iterator() for ArrayList returns a new one:
ArrayList.java:833
`public Iterator<E> iterator() {
      return new Itr();
    }
`

**Comment by haraldrombaut** [comment]
That's just the way mongodb works. This is no different behavior from the mongodb java driver. A workaround is to first create a list of the iterator and then iterate the list. Take into account the warning on http://api.mongodb.org/java/current/com/mongodb/DBCursor.html:

Warning: Calling toArray or length on a DBCursor will irrevocably turn it into an array. This means that, if the cursor was iterating over ten million results (which it was lazily fetching from the database), suddenly there will be a ten-million element array in memory. Before converting to an array, make sure that there are a reasonable number of results using skip() and limit().

**Comment by eric-2d** [comment]
Nah maybe I wasn't clear. The problem is completely unrelated to MongoDB and is about how Java iterators are supposed to work.

When you call `iterator()` you are supposed to get a `new` one. Jongo implementation returns the current iterator:
Aggregate.java:85
```  
    public Iterator<E> iterator() {
      return this;
    }
```


So if you really want to here is Mongo java driver vs Jongo:
```
// Mongo Java Driver
        FindIterable<Document> cursor = collection.find();
        MongoCursor<Document>  it     = cursor.iterator();
        while (it.hasNext()) {
            // do something
            it.next();
        }

        MongoCursor<Document> it2 = cursor.iterator();
        while (it2.hasNext()) {
            // do something else
            System.out.println("yay! I can go over my collection again cuz I requested a new iterator()");
            it2.next();
        }


// Jongo

        org.jongo.MongoCollection rawFutures = jongo.getCollection(collectionName);
        Iterable<MyObject> res = rawFutures .aggregate("").as(MyObject.class);

        Iterator<MyObject> jit = res.iterator();
        while (jit.hasNext()) {
            // do something
            System.out.println(jit.next());
        }
        
        Iterator<MyObject> jit2 = res.iterator();
        while (jit2.hasNext()) {
            // do something else
            // Well no such luck cuz jit2 == jit which already at the end of the collection.
           jit2.next();
        }
```

So as a comparison here Mongo's implementation of `iterator()`:
DBCursor.java:630
```
    @Override
    public Iterator<DBObject> iterator() {
        return this.copy(); // Which in turns does new();
    }
```

Bottom line: ``ResultIterator<E>::iterator()`` should return a new iterator not itself.

**Comment by haraldrombaut** [comment]
Well, requesting the iterator again (```DBCursor::copy()```), does cause the query to be executed again. Using the mongo java driver, this is not transparent. Using jongo on the other had it is (although not very well documented).

I don't know if this was a design decision, but it does make it more clear that the query is executed each time you iterate the resultset.
